test/fuzz/lua/test_engine.lua

```diff
@@ -0,0 +1,1411 @@
+    space:update(key, tuple_ops)
+end
+
+local function replace_op(space, tuple)
+    space:replace(tuple)
+end
+
+local function bsize_op(space)
+    space:bsize()
+end
+
+local function len_op(space)
+    space:len()
+end
+
+local function format_op(space, space_format)
+    space:format(space_format)
+end
+
+local function setup(engine_name, space_id_func, test_dir, verbose)
+    log.info('SETUP')
+    assert(engine_name == 'memtx' or
+           engine_name == 'vinyl')
+    -- Configuration reference (box.cfg),
+    -- https://www.tarantool.io/en/doc/latest/reference/configuration/
+    local box_cfg_options = {
+        checkpoint_count = math.random(5),
+        checkpoint_interval = math.random(60),
+        checkpoint_wal_threshold = math.random(1024),
+        iproto_threads = math.random(1, 10),
+        memtx_allocator = oneof({'system', 'small'}),
+        memtx_memory = 1024 * 1024,
+        memtx_sort_threads = math.random(1, 256),
+        memtx_use_mvcc_engine = oneof({true, false}),
+        readahead = 16320,
+        slab_alloc_factor = math.random(1, 2),
+        vinyl_bloom_fpr = math.random(50) / 100,
+        vinyl_cache = oneof({0, 2}) * 1024 * 1024,
+        vinyl_max_tuple_size = math.random(0, 100000),
+        vinyl_memory = 800 * 1024 * 1024,
+        vinyl_page_size = math.random(1024, 2048),
+        vinyl_range_size = 128 * 1024,
+        vinyl_read_threads = math.random(2, 10),
+        vinyl_run_count_per_level = math.random(1, 10),
+        vinyl_run_size_ratio = math.random(2, 5),
+        vinyl_timeout = math.random(1, 5),
+        vinyl_write_threads = math.random(2, 10),
+        wal_cleanup_delay = 14400,
+        wal_dir_rescan_delay = math.random(1, 20),
+        wal_max_size = math.random(1024 * 1024 * 1024),
+        wal_mode = oneof({'write', 'fsync'}),
+        wal_queue_max_size = 16777216,
+        work_dir = test_dir,
+        worker_pool_threads = math.random(1, 10),
+    }
+    if verbose then
+        box_cfg_options.log_level = 'verbose'
+    end
+    box.cfg(box_cfg_options)
+    log.info('FINISH BOX.CFG')
+
+    log.info('CREATE A SPACE')
+    local space_format = random_space_format()
+    -- TODO: support `constraint`.
+    -- TODO: support `foreign_key`.
+    local space_opts = {
+        engine = engine_name,
+        field_count = oneof({0, table.getn(space_format)}),
+        format = space_format,
+        if_not_exists = oneof({true, false}),
+        is_local = oneof({true, false}),
+    }
+    if space_opts.engine ~= 'vinyl' then
+        space_opts.temporary = oneof({true, false})
+    end
+    local space_name = ('test_%d'):format(space_id_func())
+    local space = box.schema.space.create(space_name, space_opts)
+    index_create_op(space)
+    index_create_op(space)
+    log.info('FINISH SETUP')
+    return space
+end
+
+local function teardown(space, dir)
+    log.info('TEARDOWN')
+    space:drop()
+    if dir ~= nil then
+        rmtree(dir)
```

Comment:
Removing files from under Tarantool may result in unpredictable behavior. Let's leave the directory be. Also, let's add an option for the work dir (current directory by default).

Reply:
With a patch below by default, test uses directory `test_engine_dir` in a current directory. Custom test directory can be specified with option `--test_dir`. The script will clean up the directory before testing if it exists.

```diff
diff --git a/test/fuzz/lua/test_engine.lua b/test/fuzz/lua/test_engine.lua
index 995ef2d3c9..aba1724edd 100644
--- a/test/fuzz/lua/test_engine.lua
+++ b/test/fuzz/lua/test_engine.lua
@@ -19,11 +19,15 @@ local datetime = require('datetime')
 local decimal = require('decimal')
 local uuid = require('uuid')
 
+local test_dir_name = 'test_engine_dir'
+local DEFAULT_TEST_DIR = fio.pathjoin(fio.cwd(), test_dir_name)
+
 local params = require('internal.argparse').parse(arg, {
     { 'engine', 'string' },
     { 'h', 'boolean' },
     { 'seed', 'number' },
     { 'test_duration', 'number' },
+    { 'test_dir', 'string' },
     { 'verbose', 'boolean' },
     { 'workers', 'number' },
 })
@@ -51,20 +55,21 @@ end
 local index_id_func = counter()
 
 if params.help or params.h then
-    print([[
+    print(([[
 
  Usage: tarantool test_engine.lua [options]
 
  Options can be used with '--', followed by the value if it's not
  a boolean option. The options list with default values:
 
-   workers <number, 50>          - number of fibers to run simultaneously
-   test_duration <number, 2*60>  - test duration time (sec)
-   engine <string, 'vinyl'>      - engine ('vinyl', 'memtx')
-   seed <number>                 - set a PRNG seed
-   verbose <boolean, false>      - enable verbose logging
-   help (same as -h)             - print this message
-]])
+   workers <number, 50>                  - number of fibers to run simultaneously
+   test_duration <number, 2*60>          - test duration time (sec)
+   test_dir <string, ./%s>  - path to a test directory
+   engine <string, 'vinyl'>              - engine ('vinyl', 'memtx')
+   seed <number>                         - set a PRNG seed
+   verbose <boolean, false>              - enable verbose logging
+   help (same as -h)                     - print this message
+]]):format(test_dir_name))
     os.exit(0)
 end
 
@@ -74,6 +79,9 @@ local arg_num_workers = params.workers or 50
 -- Test duration time.
 local arg_test_duration = params.test_duration or 2*60
 
+-- Test directory.
+local arg_test_dir = params.test_dir or DEFAULT_TEST_DIR
+
 -- Tarantool engine.
 local arg_engine = params.engine or 'vinyl'
 
@@ -125,15 +133,15 @@ local function keys(t)
     return table_keys
 end
 
-local function rmtree(s)
-    log.info(('CLEANUP %s'):format(s))
-    if (fio.path.is_file(s) or fio.path.is_link(s)) then
-        fio.unlink(s)
+local function rmtree(path)
+    log.info(('CLEANUP %s'):format(path))
+    if (fio.path.is_file(path) or fio.path.is_link(path)) then
+        fio.unlink(path)
         return
     end
-    if fio.path.is_dir(s) then
-        for _, i in pairs(fio.listdir(s)) do
-            rmtree(s..'/'..i)
+    if fio.path.is_dir(path) then
+        for _, p in pairs(fio.listdir(path)) do
+            rmtree(fio.pathjoin(path, p))
         end
     end
 end
@@ -543,15 +551,19 @@ local function setup(engine_name, space_id_func, test_dir, verbose)
     return space
 end
 
-local function teardown(space, dir)
-    log.info('TEARDOWN')
-    space:drop()
+local function cleanup_dir(dir)
+    log.info('CLEANUP')
     if dir ~= nil then
         rmtree(dir)
         dir = nil -- luacheck: ignore
     end
 end
 
+local function teardown(space)
+    log.info('TEARDOWN')
+    space:drop()
+end
+
 -- Indexes,
 -- https://www.tarantool.io/en/doc/latest/concepts/data_model/indexes/
 -- space_object:create_index(),
@@ -1358,11 +1370,17 @@ local function process_errors(error_messages)
     return found_unexpected_errors
 end
 
-local function run_test(num_workers, test_duration, engine_name, verbose_mode)
-    local workers = {}
+local function run_test(num_workers, test_duration, test_dir,
+                        engine_name, verbose_mode)
+
+    if fio.path.exists(test_dir) then
+        cleanup_dir(test_dir)
+    else
+        fio.mkdir(test_dir)
+    end
 
+    local workers = {}
     local space_id_func = counter()
-    local test_dir = fio.tempdir()
     local space = setup(engine_name, space_id_func, test_dir, verbose_mode)
 
     local test_gen = fun.cycle(fun.iter(keys(ops)))
@@ -1405,10 +1423,11 @@ local function run_test(num_workers, test_duration, engine_name, verbose_mode)
         end
     end
 
-    teardown(space, test_dir)
+    teardown(space)
 
     local exit_code = process_errors(error_messages) and 1 or 0
     os.exit(exit_code)
 end
 
-run_test(arg_num_workers, arg_test_duration, arg_engine, arg_verbose)
+run_test(arg_num_workers, arg_test_duration, arg_test_dir,
+         arg_engine, arg_verbose)
```